assembly/src/parsers/u32_ops.rs

```diff
@@ -411,102 +411,198 @@ pub fn parse_u32not(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
+                span_ops.push(Operation::Push(Felt::new(2u64.pow(x))));
+                true
+            }
+        },
         _ => return Err(AssemblyError::extra_param(op)),
+    };
+
+    span_ops.push(Operation::U32mul);
+
+    if drop_remainder {
+        span_ops.push(Operation::Drop);
     }
 
     Ok(())
 }
 
-/// Translates u32shr.x assembly instruction to VM operations.
+/// Translates u32shr assembly instruction to VM operations.
 ///
-/// The base operation is implemented as `PUSH(2^x) U32DIV SWAP DROP`, and we also need to make
-/// sure that the top of the stack is a u32 value.
+/// The safe modes of the operation are implemented by putting a power of 2 on the stack, then
+/// dividing the value to be shifted by it and returning the quotient. For unsafe mode, a left shift
+/// is implemented via multiplication and both the shifted value and the overflow shift are
+/// returned. For safe variations, the shift value is asserted to be between 0-31 and the value to
+/// be shifted is asserted to be a 32-bit value.
 ///
-/// With the u32 checks included, the total number of VM cycles needed is 7.
+/// VM cycles per mode:
+/// - u32shr: 11 cycles
+/// - u32shr.b: 6 cycles
+/// - u32shr.unsafe: 7 cycles
 pub fn parse_u32shr(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
-    match op.num_parts() {
-        0..=1 => return Err(AssemblyError::missing_param(op)),
-        2 => {
-            // assert the top of the stack is a u32 value
-            assert_u32(span_ops);
-
-            let x = parse_int_param(op, 1, 0, 31)?;
-            span_ops.push(Operation::Push(Felt::new(2u64.pow(x))));
-            span_ops.push(Operation::U32div);
-            // drop the remainder and keep the quotient
-            span_ops.push(Operation::Drop);
+    let drop_remainder = match op.num_parts() {
+        0 => return Err(AssemblyError::invalid_op(op)),
+        1 => {
+            // Assume the dynamic shift value b is on top of the stack.
+            span_ops.push(Operation::Pow2);
+            assert_u32_operands(span_ops, true);
+            true
         }
+        2 => match op.parts()[1] {
+            "unsafe" => {
+                span_ops.push(Operation::Push(Felt::new(32)));
+                span_ops.push(Operation::Swap);
+                span_ops.push(Operation::U32sub);
+                span_ops.push(Operation::Drop);
+                span_ops.push(Operation::Pow2);
+                span_ops.push(Operation::U32mul);
+                span_ops.push(Operation::Swap);
+                false
+            }
+            _ => {
+                assert_u32(span_ops);
+
+                let x = parse_int_param(op, 1, 0, 31)?;
+                span_ops.push(Operation::Push(Felt::new(2u64.pow(x))));
+                true
+            }
+        },
         _ => return Err(AssemblyError::extra_param(op)),
+    };
+
+    if drop_remainder {
+        span_ops.push(Operation::U32div);
```

Comment:
At first I thought this was a mistake - so, maybe add a comment about why the division is happening here? My understanding is that for the `unsafe` case, we compute everything in the `match` above, but for other cases, the actual "shift" is happening here.

Reply:
Yep, I've added comments in both places.